File: src/Patterns/MVC/NotMVC/Flock.java
package Patterns.MVC.NotMVC;

import java.util.ArrayList;
import java.util.Iterator;

/**
 * Created by Home on 11.12.2016.
 */
public class Flock implements Quackable {
    ArrayList quackers = new ArrayList();

    public void add(Quackable quacker){
        quackers.add(quacker);
    }

    @Override
    public void quack() {
        Iterator iterator = quackers.iterator();
        while (iterator.hasNext()){
            Quackable quacker = (Quackable)iterator.next();
            quacker.quack();
        }
    }

    @Override
    public void registerObserver(Observer observer) {
        Iterator iterator = quackers.iterator();
        while (iterator.hasNext()){
            Quackable quacker = (Quackable)iterator.next();
            quacker.registerObserver(observer);
        }
    }

    @Override
    public void notifyObservers() {
    }
}

File: src/patterns/together/composite/Flock.kt
package patterns.together.composite

import patterns.together.Quackable
import patterns.together.observer.Observer

class Flock : Quackable {
    val quackers = arrayListOf<Quackable>()

    override fun registerObserver(observer: Observer) {
        val iterator = quackers.iterator()
        while (iterator.hasNext()) {
            val quacker = iterator.next()
            quacker.registerObserver(observer)
        }
    }

    override fun notifyObservers() {}

    fun add(quacker: Quackable) = quackers.add(quacker)

    override fun quack() {
        val iterator = quackers.iterator()
        while (iterator.hasNext()) {
            val quacker = iterator.next()
            quacker.quack()
        }
    }
}